src/peep/peep.c

```diff
@@ -7772,13 +7772,370 @@ static void peep_on_enter_or_exit_ride(rct_peep *peep, int rideIndex, int flags)
+			return 0;
+	}
+
+///loc_69B03B:
+	if (shop_item_is_food(shopItem) && (peep->hunger > 75)) {
+		peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_NOT_HUNGRY, 0xFF); // Null arg?
+		return 0;
+	}
+
+///loc_69B069:
+	if (shop_item_is_drink(shopItem) && (peep->thirst > 75)) {
+		peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_NOT_THIRSTY, 0xFF);
+		return 0;
+	}
+
+///loc_69B097:
+	if ((shopItem == SHOP_ITEM_UMBRELLA) && (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_RAIN_LEVEL, sint8) != 0))
+			goto loc_69B119;
+
+///loc_69B0A4:
+	if ((shopItem == SHOP_ITEM_MAP) && (peep->guest_heading_to_ride_id == 0xFF))
+			goto loc_69B119;
+
+///loc_69B0B1:
+	switch (shopItem) {
+	case SHOP_ITEM_BALLOON:
+	case SHOP_ITEM_TOY:
+	case SHOP_ITEM_PHOTO:
+	case SHOP_ITEM_UMBRELLA:
+	case SHOP_ITEM_HAT:
+	case SHOP_ITEM_TSHIRT:
+	case SHOP_ITEM_PHOTO2:
+	case SHOP_ITEM_PHOTO3:
+	case SHOP_ITEM_PHOTO4:
+	case SHOP_ITEM_SUNGLASSES:
+		if (!has_voucher) {
+			if (((scenario_rand() & 0x7F) + 0x73) > peep->happiness)
+				return 0;
+			else if (peep->no_of_rides < 3)
+				return 0;
+		}
+		break;
+	default:
+		break;
+	}
+
+loc_69B119:
+	if (!has_voucher) {
+		if (price != 0) {
+			if (peep->cash_in_pocket == 0) {
+				peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_SPENT_MONEY, 0xFF);
+				return 0;
+			}
+			if (price > peep->cash_in_pocket) {
+				peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_CANT_AFFORD, shopItem);
+				return 0;
+			}
+		}
+
+		///loc_69B143:
+		// There are six tangentally-related array accesses in this chunk of code.
+		// Each three are related to each other, so I think there's two arrays.
+		// I think these arrays are arrays of structs due to the way they're accessed.
+		// The pointer base is incremented by 2 as you go down.
+		// But what else accesses this memory?
+		// Perhaps it's some kind of table?
+		// TODO: Make sure the stride is correct (it's the "* 4" part)
+		if (shopItem >= 32) { // Right side
+			// TODO: Make this a switch on refactor
+			value = RCT2_ADDRESS(0x00982246, sint16)[shopItem2 * 4];
+			if (RCT2_ADDRESS_CURRENT_TEMPERATURE >= 21)
+				value = RCT2_ADDRESS(0x00982248, sint16)[shopItem2 * 4];
+			if (RCT2_ADDRESS_CURRENT_TEMPERATURE <= 11)
+				value = RCT2_ADDRESS(0x0098224A, sint16)[shopItem2 * 4];
+		}
+		else { // Left side
+			value = RCT2_ADDRESS(0x00982166, sint16)[shopItem2 * 4];
+			if (RCT2_ADDRESS_CURRENT_TEMPERATURE >= 21)
+				value = RCT2_ADDRESS(0x00982168, sint16)[shopItem2 * 4];
+			if (RCT2_ADDRESS_CURRENT_TEMPERATURE <= 11)
+				value = RCT2_ADDRESS(0x0098216A, sint16)[shopItem2 * 4];
+		}
+
+		///loc_69B19D:
+		// Looks like this is logic for determining value
+		// i.e whether they're willing to pay the asking price
+		if (value < price) {
+			value -= price;
+			if (shopItem == SHOP_ITEM_UMBRELLA) {
+				if (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_RAIN_LEVEL, sint8) != 0)
+					goto loc_69B221;
+			}
+			///loc_69B1F8:
+			value = -value;
+			if (peep->happiness >= 128)
+				value >>= 1;
```

Comment:
If the value you are dividing is unsigned, it will optimise to a single shift right. If its signed then it will do that plus a bit fix for negative numbers. Its worth know this.